### DISCUSSION TITLE
Protobuf -> Open Api Schemas

### DISCUSSION THREAD

**Discussion by sam-heilbron** [cuelang/cue#917] [Q&A]
Is there a way to skip a particular proto message from generating an open api schema?

**Comment by myitcv** [comment]
Apologies for the delay in replying to your question, @sam-heilbron!

I'm not aware of any construct within the protocol buffers spec that would allow us to programmatically determine whether a message should be considered non-exported or not (assuming those are the semantics you are implying here?)

Did you have a particular mechanism in mind? 

The best approach to my mind would be to first import the proto to CUE, then manipulate the CUE AST to remove definitions you do not want, then generate OpenAPI.

**Reply by myitcv** [reply]
Quick demo of what that might look like:

```
exec go mod tidy
exec go run main.go

-- go.mod --
module mod.com

go 1.16

require cuelang.org/go v0.4.0
-- main.go --
package main

import (
	"fmt"
	"log"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/ast"
	"cuelang.org/go/encoding/openapi"
	"cuelang.org/go/encoding/protobuf"
)

func main() {
	f, err := protobuf.Extract("x.proto", nil, nil)
	if err != nil {
		log.Fatal(err)
	}
	decls := f.Decls
	f.Decls = nil
	for _, d := range decls {
		switch d := d.(type) {
		case *ast.Field:
			l, _, err := ast.LabelName(d.Label)
			if err != nil {
				log.Fatal(err)
			}
			if l == "#M2" {
				continue
			}
		}
		f.Decls = append(f.Decls, d)
	}
	// Note: using cue.Runtime because of cuelang.org/issue/1015
	var r cue.Runtime
	inst, err := r.CompileFile(f)
	if err != nil {
		log.Fatal(err)
	}
	byts, err := openapi.Gen(inst, nil)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("%s", byts)
}
-- x.cue --
#M1: {
	Name?: string @protobuf(1,string)
}

#M2: {
	Name?: string @protobuf(1,string)
}
-- x.proto --

message M1 { string Name = 1; }

message M2 { string Name = 1; }
```

Gives:

```
> exec go run main.go
[stdout]
{"openapi":"3.0.0","info":{"title":"Generated by cue.","version":"no version"},"paths":{},"components":{"schemas":{"M1":{"type":"object","properties":{"Name":{"type":"string"}}}}}}
```

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/917.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
